src/main/java/com/google/cloud/tools/app/impl/cloudsdk/CloudSdkCli.java

```diff
@@ -0,0 +1,261 @@
+      String gcloudOutputFormat, Map<String, String> gcloudEnvironment) {
+    this.sdk = sdk;
+    this.processRunner = processRunner;
+    this.gcloudCredentialOverride = gcloudCredentialOverride;
+    this.gcloudOutputFormat = gcloudOutputFormat;
+    this.gcloudEnvironment = gcloudEnvironment;
+    this.isUsed = false;
+  }
+
+  public enum Environment {
+    CLOUDSDK_METRICS_ENVIROMENT,
+    CLOUDSDK_METRICS_ENVIRONMENT_VERSION,
+    CLOUDSDK_APP_USE_GSUTIL
+  }
+
+  public static class Builder {
+
+    private CloudSdk sdk;
+    private ProcessOutputLineListener outputListener;
+    private ProcessOutputLineListener errorListener;
+    private boolean redirectError;
+    private ProcessExitListener exitListener;
+    private boolean async;
+    private Map<String, String> gcloudEnvironment = Maps.newHashMap();
+    private File gcloudCredentialOverride;
+    private String gcloudOutputFormat;
+
+    /**
+     * @param sdkPath Path to cloud sdk installation.
+     */
+    public Builder(File sdkPath) {
+      if (sdkPath == null) {
+        Path discoveredSdkPath = PathResolver.INSTANCE.getCloudSdkPath();
+        if (discoveredSdkPath == null) {
+          throw new CloudSdkConfigurationException("Google Cloud SDK path was not provided and"
+              + " could not be found in any known install locations.");
+        }
+        sdkPath = discoveredSdkPath.toFile();
+      }
+      sdk = new CloudSdk(sdkPath.toPath());
+    }
+
+    /**
+     * Run the process asynchronously.
+     */
+    public Builder async() {
+      this.async = true;
+      return this;
+    }
+
+    /**
+     * Client listener on the process output stream.
+     */
+    public Builder stdOutListener(ProcessOutputLineListener listener) {
+      this.outputListener = listener;
+      return this;
+    }
+
+    /**
+     * Client listener on the process error stream.
+     */
+    public Builder stdErrListener(ProcessOutputLineListener listener) {
+      this.errorListener = listener;
+      return this;
+    }
+
+    /**
+     * Client listener of the process exit with code.
+     */
+    public Builder exitListener(ProcessExitListener listener) {
+      this.exitListener = listener;
+      return this;
+    }
+
+    /**
+     * Redirect error stream to standard output stream.
+     */
+    public Builder redirectError(boolean redirect) {
+      this.redirectError = redirect;
+      return this;
+    }
+
+    /**
+     * Override file for gcloud credential configuration.
+     */
+    public Builder gcloudCredentialOverride(File override) {
+      this.gcloudCredentialOverride = override;
+      return this;
+    }
+
+    /**
+     * Format for gcloud output.
+     */
+    public Builder gcloudOutputFormat(String format) {
+      this.gcloudOutputFormat = format;
+      return this;
+    }
+
+    /**
+     * An environment variable for gcloud invocations.
+     */
+    public Builder environmentVariable(Environment key, String value) {
```

Comment:
I wonder if there's any value in hiding the environment variables from the client too. I.e. pass them in like any other gcloud parameter so the client doesn't need to know if its a flag or env var.